### PR TITLE
[Merged by Bors] - Feat: port GroupTheory.Subsemigroup.Basic

### PR DESCRIPTION
**Pull request by urkud**
.

### CODE REVIEW

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+# Subsemigroups: definition and `complete_lattice` structure
+
+This file defines bundled multiplicative and additive subsemigroups. We also define
+a `complete_lattice` structure on `Subsemigroup`s,
+and define the closure of a set as the minimal subsemigroup that includes this set.
+
+## Main definitions
+
+* `Subsemigroup M`: the type of bundled subsemigroup of a magma `M`; the underlying set is given in
+  the `carrier` field of the structure, and should be accessed through coercion as in `(S : set M)`.
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:26]:
```suggestion
  the `carrier` field of the structure, and should be accessed through coercion as in `(S : Set M)`.
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+* `AddSubsemigroup M` : the type of bundled subsemigroups of an additive magma `M`.
+
+For each of the following definitions in the `Subsemigroup` namespace, there is a corresponding
+definition in the `AddSubsemigroup` namespace.
+
+* `Subsemigroup.copy` : copy of a subsemigroup with `carrier` replaced by a set that is equal but
+  possibly not definitionally equal to the carrier of the original `Subsemigroup`.
+* `Subsemigroup.closure` :  semigroup closure of a set, i.e.,
+  the least subsemigroup that includes the set.
+* `Subsemigroup.gi` : `closure : set M → Subsemigroup M` and coercion `coe : Subsemigroup M → set M`
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:36]:
```suggestion
* `Subsemigroup.gi` : `closure : Set M → Subsemigroup M` and coercion `coe : Subsemigroup M → Set M`
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+  ⟨fun s =>
+    { carrier := ⋂ t ∈ s, ↑t
+      mul_mem' := fun hx hy =>
+        Set.mem_binterᵢ fun i h =>
+          i.mul_mem (by apply Set.mem_interᵢ₂.1 hx i h) (by apply Set.mem_interᵢ₂.1 hy i h) }⟩
+
+@[simp, norm_cast, to_additive]
+theorem coe_Inf (S : Set (Subsemigroup M)) : ((infₛ S : Subsemigroup M) : Set M) = ⋂ s ∈ S, ↑s :=
+  rfl
+#align subsemigroup.coe_Inf Subsemigroup.coe_Inf
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:238]:
```suggestion
theorem coe_infₛ (S : Set (Subsemigroup M)) : ((infₛ S : Subsemigroup M) : Set M) = ⋂ s ∈ S, ↑s :=
  rfl
#align subsemigroup.coe_Inf Subsemigroup.coe_infₛ
```

Comment (dupuisf) [reply Mathlib/GroupTheory/Subsemigroup/Basic.lean:238]: and so on for `Inf`

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+
+@[to_additive]
+theorem mem_Inf {S : Set (Subsemigroup M)} {x : M} : x ∈ infₛ S ↔ ∀ p ∈ S, x ∈ p :=
+  Set.mem_interᵢ₂
+#align subsemigroup.mem_Inf Subsemigroup.mem_Inf
+
+@[to_additive]
+theorem mem_infi {ι : Sort _} {S : ι → Subsemigroup M} {x : M} : (x ∈ ⨅ i, S i) ↔ ∀ i, x ∈ S i := by
+  simp only [infᵢ, mem_Inf, Set.forall_range_iff]
+#align subsemigroup.mem_infi Subsemigroup.mem_infi
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:248]:
```suggestion
theorem mem_infᵢ {ι : Sort _} {S : ι → Subsemigroup M} {x : M} : (x ∈ ⨅ i, S i) ↔ ∀ i, x ∈ S i := by
  simp only [infᵢ, mem_Inf, Set.forall_range_iff]
#align subsemigroup.mem_infi Subsemigroup.mem_infᵢ
```

Comment (dupuisf) [reply Mathlib/GroupTheory/Subsemigroup/Basic.lean:248]: And the same throughout. Also, `Sup` becomes `supₛ` and `supr` becomes `supᵢ`.

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+
+@[to_additive]
+theorem closure_union (s t : Set M) : closure (s ∪ t) = closure s ⊔ closure t :=
+  (Subsemigroup.gi M).gc.l_sup
+#align subsemigroup.closure_union Subsemigroup.closure_union
+
+@[to_additive]
+theorem closure_Union {ι} (s : ι → Set M) : closure (⋃ i, s i) = ⨆ i, closure (s i) :=
+  (Subsemigroup.gi M).gc.l_supᵢ
+#align subsemigroup.closure_Union Subsemigroup.closure_Union
```
Comment (dupuisf) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:413]:
```suggestion
theorem closure_unionᵢ {ι} (s : ι → Set M) : closure (⋃ i, s i) = ⨆ i, closure (s i) :=
  (Subsemigroup.gi M).gc.l_supᵢ
#align subsemigroup.closure_Union Subsemigroup.closure_unionᵢ 
```
And likewise `Inter` becomes `interᵢ`

**Comment by jcommelin** [comment]
@urkud I think the question in your top post is now answered, right?

**Comment by urkud** [comment]
> @urkud I think the question in your top post is now answered, right?

Yes.

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+end NonAssoc
+
+section Assoc
+
+namespace MulHom
+
+open Subsemigroup
+
+/-- Let `s` be a subset of a semigroup `M` such that the closure of `s` is the whole semigroup.
+Then `mul_hom.of_mdense` defines a mul homomorphism from `M` asking for a proof
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:479]:
```suggestion
Then `MulHom.ofDense` defines a mul homomorphism from `M` asking for a proof
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+    (hmul : ∀ (x), ∀ y ∈ s, f (x * y) = f x * f y) :
+    M →ₙ* N where 
+  toFun := f
+  map_mul' x y :=
+    dense_induction y hs (fun y hy x => hmul x y hy)
+      (fun y₁ y₂ h₁ h₂ x => by simp only [← mul_assoc, h₁, h₂]) x
+#align mul_hom.of_mdense MulHom.ofDense
+
+/-- Let `s` be a subset of an additive semigroup `M` such that the closure of `s` is the whole
+semigroup.  Then `add_hom.of_mdense` defines an additive homomorphism from `M` asking for a proof
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:492]:
```suggestion
semigroup.  Then `AddHom.ofDense` defines an additive homomorphism from `M` asking for a proof
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+  predicates with two arguments."]
+theorem closure_induction₂ {p : M → M → Prop} {x} {y : M} (hx : x ∈ closure s) (hy : y ∈ closure s)
+    (Hs : ∀ x ∈ s, ∀ y ∈ s, p x y) (Hmul_left : ∀ x y z, p x z → p y z → p (x * y) z)
+    (Hmul_right : ∀ x y z, p z x → p z y → p z (x * y)) : p x y :=
+  closure_induction hx
+    (fun x xs => closure_induction hy (Hs x xs) fun z _ h₁ h₂ => Hmul_right z _ _ h₁ h₂)
+    fun _ _ h₁ h₂ => Hmul_left _ _ _ h₁ h₂
+#align subsemigroup.closure_induction₂ Subsemigroup.closure_induction₂
+
+/-- If `s` is a dense set in a magma `M`, `subsemigroup.closure s = ⊤`, then in order to prove that
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:366]:
```suggestion
/-- If `s` is a dense set in a magma `M`, `Subsemigroup.closure s = ⊤`, then in order to prove that
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+some predicate `p` holds for all `x : M` it suffices to verify `p x` for `x ∈ s`,
+and verify that `p x` and `p y` imply `p (x * y)`. -/
+@[elab_as_elim, to_additive "If `s` is a dense set in an additive monoid `M`,
+  `add_subsemigroup.closure s = ⊤`, then in order to prove that some predicate `p` holds
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:370]:
```suggestion
  `AddSubsemigroup.closure s = ⊤`, then in order to prove that some predicate `p` holds
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+  ⟨Subsemigroup.carrier, fun p q h => by cases p; cases q; congr ⟩
+
+@[to_additive]
+instance : MulMemClass (Subsemigroup M) M where mul_mem := fun {_ _ _} => Subsemigroup.mul_mem' _
+
+/-- See Note [custom simps projection] -/
+@[to_additive " See Note [custom simps projection]"]
+def Simps.coe (S : Subsemigroup M) : Set M :=
+  S
+#align subsemigroup.simps.coe Subsemigroup.Simps.coe
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:114]: I'm afraid that you need to manually add `#align` statements for the to_additive versions...

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+@[elab_as_elim, to_additive "An induction principle for additive closure membership. If `p`
+  holds for all elements of `s`, and is preserved under addition, then `p` holds for all
+  elements of the additive closure of `s`."]
+theorem closure_induction {p : M → Prop} {x} (h : x ∈ closure s) (Hs : ∀ x ∈ s, p x)
+    (Hmul : ∀ x y, p x → p y → p (x * y)) : p x :=
+  (@closure_le _ _ _ ⟨p, Hmul _ _⟩).2 Hs h
+#align subsemigroup.closure_induction Subsemigroup.closure_induction
+#align add_subsemigroup.closure_induction AddSubsemigroup.closure_induction
+
+/-- A dependent version of `subsemigroup.closure_induction`.  -/
```
Comment (ChrisHughes24) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:371]:
```suggestion
/-- A dependent version of `Subsemigroup.closure_induction`.  -/
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+@[elab_as_elim, to_additive "A dependent version of `add_subsemigroup.closure_induction`. "]
```
Comment (ChrisHughes24) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:372]:
```suggestion
@[elab_as_elim, to_additive "A dependent version of `AddSubsemigroup.closure_induction`. "]
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+#align add_subsemigroup.closure_Union AddSubsemigroup.closure_unionᵢ 
+
+@[to_additive]
+theorem closure_singleton_le_iff_mem (m : M) (p : Subsemigroup M) : closure {m} ≤ p ↔ m ∈ p := by
+  rw [closure_le, singleton_subset_iff, SetLike.mem_coe]
+#align subsemigroup.closure_singleton_le_iff_mem Subsemigroup.closure_singleton_le_iff_mem
+#align add_subsemigroup.closure_singleton_le_iff_mem AddSubsemigroup.closure_singleton_le_iff_mem
+
+@[to_additive]
+theorem mem_supr {ι : Sort _} (p : ι → Subsemigroup M) {m : M} :
```
Comment (ChrisHughes24) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:459]:
```suggestion
theorem mem_supᵢ {ι : Sort _} (p : ι → Subsemigroup M) {m : M} :
```

File: Mathlib/GroupTheory/Subsemigroup/Basic.lean [hunk 14]
```diff
@@ -0,0 +1,550 @@
...
+  simp only [closure_singleton_le_iff_mem]
+#align subsemigroup.mem_supr Subsemigroup.mem_supr
+#align add_subsemigroup.mem_supr AddSubsemigroup.mem_supr
+
+@[to_additive]
+theorem supr_eq_closure {ι : Sort _} (p : ι → Subsemigroup M) :
+    (⨆ i, p i) = Subsemigroup.closure (⋃ i, (p i : Set M)) := by
+  simp_rw [Subsemigroup.closure_unionᵢ, Subsemigroup.closure_eq]
+#align subsemigroup.supr_eq_closure Subsemigroup.supr_eq_closure
+#align add_subsemigroup.supr_eq_closure AddSubsemigroup.supr_eq_closure
```
Comment (ChrisHughes24) [inline Mathlib/GroupTheory/Subsemigroup/Basic.lean:471]:
```suggestion
theorem supᵢ_eq_closure {ι : Sort _} (p : ι → Subsemigroup M) :
    (⨆ i, p i) = Subsemigroup.closure (⋃ i, (p i : Set M)) := by
  simp_rw [Subsemigroup.closure_unionᵢ, Subsemigroup.closure_eq]
#align subsemigroup.supr_eq_closure Subsemigroup.supᵢ_eq_closure
#align add_subsemigroup.supr_eq_closure AddSubsemigroup.supᵢ_eq_closure
```

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**